wgsl/index.bs

```diff
@@ -1368,86 +1372,384 @@ Therefore, a data value must not be placed in the padding at the end of a struct
 </table>
 
-Note: We've described a SPIR-V logical pointer type.
+<div class='example wgsl' heading='Pointer type (program fragment)'>
+  <xmp highlight='rust'>
+    // The type of a pointer value that references storage for keeping
+    // an 'i32' value, using memory locations in the 'uniform' storage class.
+    // Here 'i32' is the pointee type.
+    ptr<uniform, i32>              // Note: This is not a valid expression on its own.
+
+    // The type of a pointer value that references storage for keeping an array
+    // of 50 elements of type 'f32', using memory locations in the 'private'
+    // storage class.
+    // Here the pointee type is 'array<f32,50>'.
+    ptr<private, array<f32, 50>>   // Note: This is not a valid expression on its own.
+  </xmp>
+</div>
 
-Note: Pointers are not storable.
+Reference types and pointer types are both sets of memory views:
+a particular memory view is associated with a unique reference value and also a unique pointer value:
+
+<blockquote algorithm="pointer reference correspondence">
+Each pointer value |p| of type ptr&lt;|SC|,|T|&gt; corresponds to a unique reference value |r| of type ref&lt;|SC|,|T|&gt;,
+and vice versa,
+where |p| and |r| describe the same memory view.
+</blockquote>
+
+In [SHORTNAME] a reference value always corresponds to the memory view
+for some or all of the memory locations for some variable.
+This defines the <dfn noexport>originating variable</dfn> for the reference value.
+A pointer value always corresponds to a reference value, and so the originating variable
+of a pointer is the same as the originating variable of the corresponding reference.
+
+Note: The originating variable is a dynamic concept.
+The originating variable for a formal parameter of a function depends on the call sites for the function.
+Different call sites may supply pointers into different originating variables.
+
+References and pointers are distinguished by how they are used:
+
+* The type of a [=variable=] is a reference type.
+* The [=address-of=] operation (unary `&`) converts a reference value to its corresponding pointer value.
+* The [=indirection=] operation (unary `*`) converts a pointer value to its corresponding reference value.
+* A const declaration can be of pointer type, but not of reference type.
+* A [=formal parameter=] can be of pointer type, but not of reference type.
+* An [=assignment statement=] updates the contents of memory via a reference:
+    * The left-hand side of the assignment statement must be of reference type.
+    * The right-hand side of the assignment statement must evaluate to the store type of the left-hand side.
+* The <dfn noexport>Load Rule</dfn>: Inside a function, a reference is automatically dereferenced (read from) to satisfy type rules:
+    * In a function, when a reference expression |r| with store type |T| is used in a statement or an expression, where
+    * The only potentially matching type rules require |r| to have a value of type |T|, then
```

Comment:
This is being honest about when memory is read from a variable's storage.

We need a precise statement _like this_ so we can clearly talk about a memory model, particularly in the presence of multiple invocations executing concurrently.
Regardless of whether we support pointers or references, the spec is missing rules about what operations perform memory accesses.  It needs this level of precision.

Reply:
Right. As mentioned on the office hours call, it's great to have this explicitly stated in the spec.
What's not clear (and is concerning) is whether having it as a separate type (generic type) is the best way to express this, since it effectively becomes an implicit cast, the first of a kind in WGSL. And implicit casts are one of the things making the type system weaker. There are many sources that describe this, e.g. [this one](https://blogs.agilefaqs.com/2011/07/11/dynamic-typing-is-not-weak-typing/):
> The more type coercions (implicit conversions) for built-in operators the language offers, the weaker the typing. (This could also be viewed as more built-in overloading of built-in operators.)